capactio/capact
# Inject additional TypeInstances to Implementation via Policy

## Description

Changes proposed in this pull request:

- Inject additional TypeInstances to Implementation via Policy
- Do not allow unknown input TypeInstances during creating Action
- Update GraphQL API for Policy with new fields

**Additional:**
- Add unit tests for `PolicyEnforcedClient.ListAdditionalInputToInjectBasedOnPolicy`
- Refactor and fix Renderer test to use local Hub client along real validators

## Testing

1. Check out the branch and run:

```
DISABLE_MONITORING_INSTALLATION=true make dev-cluster
```

2. Build and use the latest CLI:

```
make build-tool-cli && mv ./bin/capact-darwin-amd64 /usr/local/bin/capact
```

3. Create fake PostgreSQL TypeInstance:

> **NOTE:** Until we don't validate the TypeInstance input, you can put values which don't match the JSON schema

```
cat > /tmp/ti.yaml << ENDOFFILE
typeInstances:
  - alias: "postgresql"
    typeRef:
      path: cap.type.database.postgresql.config
      revision: 0.1.0
    value:
      postgresql: true
ENDOFFILE

export POSTGRESQL_TI_ID=$(capact ti create --from-file /tmp/ti.yaml -ojson | jq -r '.[0].id')
```

4. Update global policy:

```bash
cat > /tmp/policy.yaml << ENDOFFILE
rules:
  - interface:
      path: "cap.interface.capactio.capact.validation.hub.install"
    oneOf:
      - implementationConstraints:
          path: "cap.implementation.capactio.capact.validation.io.install"
        inject:
          additionalParameters:
            - name: mattermost-parameters
              value:
                revisionHistoryLimit: 123
          additionalTypeInstances:
            - name: postgresql
              id: "${POSTGRESQL_TI_ID}"

  - interface:
      path: cap.*
    oneOf:
      - implementationConstraints:
          requires:
            - path: "cap.core.type.platform.kubernetes"
      - implementationConstraints: {} # fallback to any Implementation
ENDOFFILE

capact policy apply -f /tmp/policy.yaml
```

5. Create Action Policy with another fake PostgreSQL TypeInstance:

```bash
cat > /tmp/ti2.yaml << ENDOFFILE
typeInstances:
  - alias: "postgresql"
    typeRef:
      path: cap.type.database.postgresql.config
      revision: 0.1.0
    value:
      postgresql: "new"
ENDOFFILE

export NEW_POSTGRESQL_TI_ID=$(capact ti create --from-file /tmp/ti2.yaml -ojson | jq -r '.[0].id')

cat <<EOF > /tmp/act-policy.yaml
rules:
  - interface:
      path: cap.interface.capactio.capact.validation.hub.install
    oneOf:
      - implementationConstraints:
          path: "cap.implementation.capactio.capact.validation.io.install"
        inject:
          additionalTypeInstances:
            - name: postgresql
              id: "${NEW_POSTGRESQL_TI_ID}"
EOF
```

6. Create Actions

```bash
# Without Action Policy
capact act create cap.interface.capactio.capact.validation.hub.install --name test1

# With Action Policy
capact act create cap.interface.capactio.capact.validation.hub.install --name test2 --action-policy-from-file /tmp/act-policy.yaml
```

7. Get and run Actions

```
# Wait until they are ready to run
capact action get

# Check the TI are properly injected
capact action get test1 -oyaml | grep $POSTGRESQL_TI_ID
capact action get test2 -oyaml | grep $NEW_POSTGRESQL_TI_ID
# You can also check diff if you're curious 🙂
diff <(capact action get test1 -oyaml) <(capact action get test2 -oyaml) 

# Run
capact act run test1
capact act run test2
```

7. Watch Actions:
```bash
capact act watch test1 
# in a separate window
capact act watch test2
```

8. Check logs:
```bash
argo logs test1
argo logs test2
```

Verify that there's no `optional postgresql TypeInstance was not specified` output, which means the step was properly replaced with a mocked one injecting existing TypeInstance.

## Related issue(s)

https://github.com/capactio/capact/issues/438
